homeassistant/components/group/sensor.py

```diff
@@ -0,0 +1,415 @@
+        self._round_digits = round_digits
+        self._attr_state_class = state_class
+        self.calc_state_class: SensorStateClass | None = None
+        self._attr_device_class = device_class
+        self.calc_device_class: SensorDeviceClass | None = None
+        self._attr_native_unit_of_measurement = unit_of_measurement
+        self.calc_unit_of_measurement: str | None = None
+        self._attr_name = name
+        if name == DEFAULT_NAME:
+            self._attr_name = f"{DEFAULT_NAME} {sensor_type}".capitalize()
+        self._attr_extra_state_attributes = {ATTR_ENTITY_ID: entity_ids}
+        self._attr_unique_id = unique_id
+        self.mode = all if mode else any
+        self._state_calc: Callable[
+            [list[tuple[str, float]], int], tuple[dict[str, str], float | None]
+        ] = CALC_TYPES[self._sensor_type]
+        self._state_incorrect: set[str] = set()
+        self._extra_state_attribute: dict[str, Any] = {}
+
+    async def async_added_to_hass(self) -> None:
+        """Register callbacks."""
+
+        @callback
+        def async_state_changed_listener(event: Event) -> None:
+            """Handle child updates."""
+            self.async_set_context(event.context)
+            self.async_defer_or_update_ha_state()
+
+        self.async_on_remove(
+            async_track_state_change_event(
+                self.hass, self._entity_ids, async_state_changed_listener
+            )
+        )
+
+        await super().async_added_to_hass()
+
+    @callback
+    def async_update_group_state(self) -> None:
+        """Query all members and determine the sensor group state."""
+        states: list[StateType] = []
+        valid_states: list[bool] = []
+        last_updated: datetime | None = None
+        sensor_values: list[tuple[str, float]] = []
+        for entity_id in self._entity_ids:
+            if (state := self.hass.states.get(entity_id)) is not None:
+                states.append(state.state)
+                try:
+                    sensor_values.append((entity_id, float(state.state)))
+                    if entity_id in self._state_incorrect:
+                        self._state_incorrect.remove(entity_id)
+                except ValueError:
+                    valid_states.append(False)
+                    if entity_id not in self._state_incorrect:
+                        self._state_incorrect.add(entity_id)
+                        _LOGGER.warning(
+                            "Unable to use state. Only numerical states are supported, entity %s with value %s excluded from calculation",
+                            entity_id,
+                            state.state,
+                        )
+                    continue
+                valid_states.append(True)
+                if last_updated is None or state.last_updated > last_updated:
+                    last_updated = state.last_updated
+                    last = float(state.state)
+                    last_entity_id = entity_id
```

Comment:
Can we try to move this to `calc_last`?
What about if `sensor_values` is updated to `sensor_values: list[tuple[str, float, StateType]] = []`:
```py
sensor_values.append((entity_id, float(state.state), state))
```

Then `calc_last` has access to the complete state and can pick the newest one?

Reply:
Has been adapted